### ISSUE TITLE
 Nonautoregressive-Transformer performs poor

### ISSUE THREAD

**Issue by xwgeng** [pytorch/fairseq#1316]
Thank you for publishing the code of NAT. I tried to reproduce the results of [vanilla NAT](https://arxiv.org/pdf/1711.02281.pdf) following the [instruction](https://github.com/pytorch/fairseq/blob/master/examples/nonautoregressive_translation/scripts.md). I trained NAT on [distilled WMT14 en-de dataset](http://dl.fbaipublicfiles.com/nat/distill_dataset.zip) provided by Fairseq with following script:
```bash
fairseq-train \
    data-bin/wmt14_en_de_distill \
    --save-dir checkpoints \
    --ddp-backend=no_c10d \
    --task translation_lev \
    --criterion nat_loss \
    --arch nonautoregressive_transformer \
    --noise full_mask \
    --share-all-embeddings \
    --optimizer adam --adam-betas '(0.9,0.98)'  \
    --lr 0.0005 --lr-scheduler inverse_sqrt \
    --min-lr '1e-09' --warmup-updates 10000 \
    --warmup-init-lr '1e-07' --label-smoothing 0.1 \
    --dropout 0.3 --weight-decay 0.01 \
    --decoder-learned-pos \
    --encoder-learned-pos \
    --pred-length-offset \
    --length-loss-factor 0.1 \
    --apply-bert-init \
    --log-format 'simple' --log-interval 100 \
    --fixed-validation-seed 7 \
    --max-tokens 8000 \
    --save-interval-updates 10000 \
    --max-update 300000
```
I tested the performance of NAT with following script:
```bash
python generate.py 
   data-bin/wmt14_en_de/ 
--gen-subset test 
--task translation_lev 
--path ../checkpoints/Non-autoregressive_Transformer/checkpoint_best.pt 
--iter-decode-max-iter 0
--beam 1 --remove-bpe 
--print-step 
--batch-size 400
```
Besides, I also attempted to incorporate copying source embedding, positional attention and layer-wise attention. However, BLEU on test set (newstest2014) is about 7 BLEU, while it's about 18 BLEU on development set (newstest2013). 

I have two following questions about training NAT:
1. why does a large margin between Fairseq's implementation(about 7 BLEU) and [vanilla NAT](https://arxiv.org/pdf/1711.02281.pdf)(17.35 BLEU) exist? Fairseq's implementation should obtain better performance due to bigger batch size and better distilled dataset generated by transformer with 27.2 BLEU, compared with vanilla NAT.
2. The trained NAT performs well on development set, but why does it has a drastic decrease by 11 BLEU on test set?

**Comment by MultiPath** [comment]
Hi,

So sorry for replying late.

1. The fairseq implementation does not use "fertility" as the latent variables in the middle, so it will get a lower BLEU score; the original paper relies on knowledge distillation. We also provided one distillation dataset, you can train on that and it will give you around 18~19 BLEU scores.
2. Yes, there is a big difference for the valid and test set in terms of BLEU scores. It is a dataset wise difference and absolute BLEU scores are not really comparable.

**Comment by xwgeng** [comment]
Thanks for your reply.

1. All above experiments were conducted on your provided distillation dataset (`http://dl.fbaipublicfiles.com/nat/distill_dataset.zip`).  I also trained another vanilla NAT using above training script, but obtaining similar results (about 7 BLEU).
2. Does you use the same training script as above, in order to obtain 18~19 BLEU scores? Whether or not incorporate source embedding copying, layer-wise attention or positional attention?

**Comment by xwgeng** [comment]
I have solved above issues. With a large batch size about 64K, vanilla NAT on distillation dataset can achieve 18.6 BLEU.

Thanks for your help. @MultiPath

**Comment by dmortem** [comment]
Hi, I met the same issue. What does "batch size about 64K" mean? Do you train on multiple GPUs and use "--max-tokens 64000"? Or train on one GPU and use "--max-tokens 8000", "--update-freq 8"? Could you provide your training script? Many thanks!

**Comment by xwgeng** [comment]
hi, @dmortem 

 ''batch size about 64k'' means a total batch size of 64k tokens, and can use single GPU with  either max-tokens as 64k and update-freq 1 or  max-tokens 8k and update-freq as 8. In my case, I trained vanilla NAT using 8 GPU with max tokens as 8k. The training script is as follows:

```bash
fairseq-train \
    data-bin/wmt14_en_de_distill \
    --save-dir checkpoints \
    --ddp-backend=no_c10d \
    --task translation_lev \
    --criterion nat_loss \
    --arch nonautoregressive_transformer \
    --noise full_mask \
    --share-all-embeddings \
    --optimizer adam --adam-betas '(0.9,0.98)'  \
    --lr 0.0005 --lr-scheduler inverse_sqrt \
    --min-lr '1e-09' --warmup-updates 10000 \
    --warmup-init-lr '1e-07' --label-smoothing 0.1 \
    --dropout 0.3 --weight-decay 0.01 \
    --decoder-learned-pos \
    --encoder-learned-pos \
    --pred-length-offset \
    --length-loss-factor 0.1 \
    --apply-bert-init \
    --log-format 'simple' --log-interval 100 \
    --fixed-validation-seed 7 \
    --max-tokens 8000 \
    --save-interval-updates 10000 \
    --max-update 300000
```

**Comment by dmortem** [comment]
Thank you! @xwgeng 

I will have a try asap. Hopefully I can get the expected result!

**Comment by hzhwcmhf** [comment]
@xwgeng I still have problems when using a large batch size....  Do we need ``--src-embedding-copy``?

**Comment by xwgeng** [comment]
@hzhwcmhf NAT can obtain expected results (about 19 BLEU) without `--src-embedding-copy`

**Comment by hzhwcmhf** [comment]
@xwgeng I tried both settings and get 18.92 BLEU without ``src-embedding-copy`` and 17.55 BLEU with ``src-embedding-copy``. 
It's really weird that the first NAT paper ("Non-Autoregressive Neural Machine Translation") mentioned that they have to copy the source embeddings or the model will be fail in training. The fairseq code authors also mention that they use copy the source embeddings in their paper ("Understanding Knowledge Distillation in Non-autoregressive Machine Translation").
Do you have any ideas about the problem?

**Comment by xwgeng** [comment]
@hzhwcmhf The reason may be that fairseq use a different and powerful NAT implementation from the original version (`https://github.com/salesforce/nonauto-nmt`) in model architecture and optimization strategies, leading to different results with `src-embedding-copy`. For examples, the original NAT must use layer-wise attention, positional attention etc. to obtain expected performance. Based on more powerful fairseq, incorporating source embedding may be not useful, as stated in `https://github.com/pytorch/fairseq/issues/1458#issuecomment-562262959`.

**Comment by hzhwcmhf** [comment]
@xwgeng It is a good explaination. Thank you very much! Maybe the author should mention it in their README and code examples.

**Comment by MultiPath** [comment]
Hi @hzhwcmhf @xwgeng Sorry for commenting on this thread late.
I think ``src-embedding-copy'' was buggy here, and the major reason is the fairseq encoder in default uses "left padding".
If you disable left-pad in the encoder side, it should be similar to standard performance.

**Comment by hzhwcmhf** [comment]
@MultiPath

Confirmed. I got very similar result after disabling left-pad (removing leftpad + src embedding copy v.s. without src embedding copy). However, it is a little different from your paper (especially from the first one). I think it can help a lot of people if you make it more clearer in ``https://github.com/pytorch/fairseq/blob/aa5f0119a383e013e56ae5d88e4a7aff0e67f0f9/examples/nonautoregressive_translation/scripts.md``.

**Comment by kkeleve** [comment]
hi, @MultiPath @xwgeng
I tried to reproduce NAT on 300k small data , and BLEU on teacher model is about 30,  but BLEU on NAT is  around 7 . Can you tell me why the performance is so poor? The script is as follows:
```
fairseq-train \
    ${databin} \
    --save-dir ${model_save} \
    --task translation_lev \
    --criterion nat_loss \
    --arch nonautoregressive_transformer  \
    --noise full_mask \
    --share-all-embeddings \
    --optimizer adam --adam-betas '(0.9, 0.98)' \
    --lr 5e-4 --lr-scheduler inverse_sqrt  \
    --stop-min-lr '1e-09' --warmup-updates 10000 \
    --warmup-init-lr '1e-07' --label-smoothing 0.1 \
    --dropout 0.3 --weight-decay 0.01 \
    --decoder-learned-pos \
    --encoder-learned-pos \
    --pred-length-offset \
    --length-loss-factor 0.1 \
    --apply-bert-init \
    --log-format  'simple' --log-interval 200 \
    --max-tokens 4096  \
    --save-interval-updates 10000 \
    --max-update 300000  \
    --keep-interval-updates 5 --keep-last-epochs 5\
    --fp16 --seed $seed  
```  
```
fairseq-generate ${data_dir} \
-s zh -t en \
--gen-subset test \
--task translation_lev \
--path ${model_save}/checkpoint_best.pt \
--batch-size 400 \
--beam 1   \
--iter-decode-max-iter 0 --print-step \
--remove-bpe
```
I don't know if batchsize or save-interval-updates is appropriate. Can you give me some advice if the hyperparameters is something wrong ?
